staging/src/k8s.io/apimachinery/pkg/util/diff/diff.go

```diff
@@ -17,122 +17,246 @@ limitations under the License.
+	if reflect.DeepEqual(a, b) {
+		return ""
+	}
 
-// StringDiff diffs a and b and returns a human readable diff.
-// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
-func StringDiff(a, b string) string {
-	return legacyDiff(a, b)
+	var sb strings.Builder
+	diffValues(&sb, reflect.ValueOf(a), reflect.ValueOf(b), "", make(map[uintptr]bool), make(map[uintptr]bool))
+	return sb.String()
 }
 
-// ObjectDiff prints the diff of two go objects and fails if the objects
-// contain unhandled unexported fields.
-// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
-func ObjectDiff(a, b interface{}) string {
-	return legacyDiff(a, b)
-}
+func diffValues(sb *strings.Builder, a, b reflect.Value, path string, visitedA, visitedB map[uintptr]bool) {
+	// Handle invalid values (nil)
+	if !a.IsValid() || !b.IsValid() {
+		if a.IsValid() != b.IsValid() {
+			formatDiff(sb, path, formatValue(a), formatValue(b))
+		}
+		return
+	}
 
-// ObjectGoPrintDiff prints the diff of two go objects and fails if the objects
-// contain unhandled unexported fields.
-// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
-func ObjectGoPrintDiff(a, b interface{}) string {
-	return legacyDiff(a, b)
-}
+	// Different types
+	if a.Type() != b.Type() {
+		formatDiff(sb, path, fmt.Sprintf("%T(%s)", a.Interface(), formatValue(a)), fmt.Sprintf("%T(%s)", b.Interface(), formatValue(b)))
+		return
+	}
 
-// ObjectReflectDiff prints the diff of two go objects and fails if the objects
-// contain unhandled unexported fields.
-// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
-func ObjectReflectDiff(a, b interface{}) string {
-	return legacyDiff(a, b)
-}
+	// Handle nil slices vs empty slices and nil maps vs empty maps
+	switch a.Kind() {
+	case reflect.Slice, reflect.Map:
+		if (a.IsNil() && !b.IsNil() && b.Len() == 0) || (!a.IsNil() && a.Len() == 0 && b.IsNil()) {
+			return
+		}
+	}
 
-// ObjectGoPrintSideBySide prints a and b as textual dumps side by side,
-// enabling easy visual scanning for mismatches.
-func ObjectGoPrintSideBySide(a, b interface{}) string {
-	sA := dump.Pretty(a)
-	sB := dump.Pretty(b)
+	// Handle based on kind
+	switch a.Kind() {
+	case reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr,
+		reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128, reflect.String:
+		if a.Interface() != b.Interface() {
+			formatDiff(sb, path, formatValue(a), formatValue(b))
+		}
 
-	linesA := strings.Split(sA, "\n")
-	linesB := strings.Split(sB, "\n")
-	width := 0
-	for _, s := range linesA {
-		l := len(s)
-		if l > width {
-			width = l
+	case reflect.Ptr, reflect.Interface:
+		if a.IsNil() || b.IsNil() {
+			if a.IsNil() != b.IsNil() {
+				formatDiff(sb, path, formatValue(a), formatValue(b))
+			}
+			return
 		}
-	}
-	for _, s := range linesB {
-		l := len(s)
-		if l > width {
-			width = l
+
+		// Check for cycles
```

Comment:
One of the nice things about rendering to JSON or YAML is that you don't need to deal with this stuff.  It doesn't handle private fields well, though.